# Assignment: Data Wrangling
### `! git clone https://github.com/ds3001f25/wrangling_assignment.git`
### Do Q1 and Q2
### Reading material: `tidy_data.pdf`

In [1]:
! git clone https://github.com/amn8ge/wrangling_assignment.git

Cloning into 'wrangling_assignment'...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 14 (delta 2), reused 0 (delta 0), pack-reused 9 (from 1)
Receiving objects: 100% (14/14), 5.83 MiB | 15.32 MiB/s, done.
Resolving deltas: 100% (2/2), done.


In [2]:
cd wrangling_assignment/

/content/wrangling_assignment


In [3]:
cd data/

/content/wrangling_assignment/data


In [4]:
ls -al

total 12192
drwxr-xr-x 2 root root    4096 Sep  1 23:43 ./
drwxr-xr-x 4 root root    4096 Sep  1 23:45 ../
-rw-r--r-- 1 root root 3414307 Sep  1 23:43 airbnb_hw.csv
-rw-r--r-- 1 root root 7181667 Sep  1 23:43 justice_data.parquet
-rw-r--r-- 1 root root 1873502 Sep  1 23:43 mn_police_use_of_force.csv


In [29]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
airbnb = pd.read_csv('airbnb_hw.csv',low_memory=False)
airbnb.head()

Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN  
3                  1                  0   199                   NaN  
4                  1                 39   549                  96.0

In [8]:
airbnb['price_clean'] = airbnb['Price'].str.replace('$', '').str.replace(',', '')
airbnb['price_clean'] = pd.to_numeric(airbnb['price_clean'], errors='coerce')
airbnb['price_clean']
airbnb.head()

# I cleaned the set by removing dollar signs and commas so that every number is
# formatted as an integer

Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  \
0                  1                  0   145                   NaN   
1                  1                  1    37                   NaN   
2                  1                  1    28                   NaN   
3                  1                  0   199                   NaN   
4                  1                 39   549                  96.0   

   price_clean  
0          145  
1           37  
2           28  
3          199  
4          549

In [15]:
missing_price_count = int(airbnb['Price'].isna().sum())
print(missing_price_count)

# there are no missing values

0


In [30]:
police = pd.read_csv('mn_police_use_of_force.csv',low_memory=False)
police.head()

response_datetime               problem is_911_call primary_offense  \
0  2016/01/01 00:47:36  Assault in Progress          Yes          DASLT1   
1  2016/01/01 02:19:34                Fight           No          DISCON   
2  2016/01/01 02:19:34                Fight           No          DISCON   
3  2016/01/01 02:28:48                Fight           No          PRIORI   
4  2016/01/01 02:28:48                Fight           No          PRIORI   

  subject_injury         force_type   force_type_action   race     sex   age  \
0            NaN       Bodily Force  Body Weight to Pin  Black    Male  20.0   
1            NaN  Chemical Irritant       Personal Mace  Black  Female  27.0   
2            NaN  Chemical Irritant       Personal Mace  White  Female  23.0   
3            NaN  Chemical Irritant  Crowd Control Mace  Black    Male  20.0   
4            NaN  Chemical Irritant  Crowd Control Mace  Black    Male  20.0   

         type_resistance  precinct   neighborhood  
0                 Tensed         1  Downtown East  
1  Verbal Non-Compliance         1  Downtown West  
2  Verbal Non-Compliance         1  Downtown West  
3    Commission of Crime         1  Downtown West  
4    Commission of Crime         1  Downtown West

In [34]:
print(police['subject_injury'].unique(),'\n')

[nan 'No' 'Yes'] 



In [18]:
police['subject_injury'].value_counts(dropna=False)

subject_injury
NaN    9848
Yes    1631
No     1446
Name: count, dtype: int64

In [27]:
missing_count = police['subject_injury'].isna().sum()
total_count = len(police)
missing_proportion = missing_count / total_count

print(missing_count)
print(missing_proportion)

9848
0.7619342359767892


76% of the values are missing, which is cause for concern.

**Q1.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing?
3. Dummy variable: For the pretrial data covered in the lecture `./data/justice_data.parquet`, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?
7. Clean the `Fatal Y/N` variable so it only takes three values: Y, N, and Unknown.
8. Are sharks more likely to launch unprovoked attacks on men or women? Is the attack more or less likely to be fatal when the attack is provoked or unprovoked? Is it more or less likely to be fatal when the victim is male or female? How do you feel about sharks?
9. What proportion of attacks appear to be by white sharks? (Hint: `str.split()` makes a vector of text values into a list of lists, split by spaces.)